File: algorithmic_toolbox/week2_algorithmic_warmup/3_greatest_common_divisor/gcd.go
package main

import (
	"C"
	"fmt"
)

func gcdNaive(a, b int) int {
	currentGcd := 1

	for d := 2; d <= a && d <= b; d++ {
		if a%d == 0 && b%d == 0 {
			if d > currentGcd {
				currentGcd = d
			}
		}
	}

	return currentGcd
}

func gcdEuclidean(a, b int) int {
	min := a
	max := b
	if a > b {
		min = b
		max = a
	}

	if min == 0 {
		return max
	}

	return gcdEuclidean(max%min, min)
}

func main() {
	var a, b int
	fmt.Scan(&a, &b)

	fmt.Println(gcdNaive(a, b))
	fmt.Println(gcdEuclidean(a, b))
}

File: algorithmic_toolbox/week2_algorithmic_warmup/3_greatest_common_divisor/GCD.java
import java.util.Scanner;

public class GCD {
  private static int gcd_naive(int a, int b) {
    int current_gcd = 1;

    for (int d = 2; d <= a && d <= b; ++d)
      if (a % d == 0 && b % d == 0)
        if (d > current_gcd)
          current_gcd = d;

    return current_gcd;
  }

  private static int gcd_euclidean(int a, int b) {
    int min = Math.min(a, b);
    int max = Math.max(a, b);

    return min == 0 ? max : gcd_euclidean(max % min, min);
  }

  public static void main(String args[]) {
    Scanner scanner = new Scanner(System.in);
    int a = scanner.nextInt();
    int b = scanner.nextInt();
    scanner.close();

    System.out.println(gcd_naive(a, b));
    System.out.println(gcd_euclidean(a, b));
  }
}
